Allow sorting of teams in a topic.

peter-fogg opened the pull request:
## [TNL-1937](https://openedx.atlassian.net/browse/TNL-1937)

Adds sorting to the teams-for-topic view in the Teams page. This reuses some of the code from #9450.

Also, let me know what you think of this PR template -- I've got a draft [here](https://openedx.atlassian.net/wiki/display/TNL/TNL+Pull+Request+Template) which I think will be a decent starting point for figuring out how to structure our code reviews.

### Sandbox
- [x] http://peter-fogg.sandbox.edx.org/courses/course-v1:TeamsX+T101+2015/teams/#my-teams

### Testing
- [x] i18n
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
Please check the appropriate box after all relevant reviewers have finished and given the :+1:.
- [ ] Code review: @andy-armstrong
- [ ] Code review: @bderusha
- [ ] Accessibility review: @cptvitamin -- for this and #9450 we've got [a story to fix any accessibility issues](https://openedx.atlassian.net/browse/TNL-1938), so you could leave any comments on that ticket.
- [ ] Product review: @explorerleslie

peter-fogg commented:
jenkins run bokchoy, or else

explorerleslie commented:
@peter-fogg default sort for teams should be last activity (right now it's alphabetical). Otherwise looks great! :+1:

bderusha commented:
Code looks good to me.  :+1:

cahrens commented:
@explorerleslie Do you even want alphabetical as a sort option? It was not specified in the JIRA ticket, and Lou didn't want it.

cahrens commented:
@bderusha pointed out that when you hover over "Sorted by", you get a hand pointer like a link.

explorerleslie commented:
@cahrens oh yeah, I forgot about that. @peter-fogg can you please remove alphabetical as a sort option.

andy-armstrong commented:
@explorerleslie @andywaldrop @peter-fogg As a nit, I find sorting by open slots to be confusing. The primary use case is to find a team that has openings, but I end up being shown all the teams that nobody is interested in first, i.e. the least popular ones. Maybe it would help to have the secondary sort be last activity so that at least I see those, but the challenge then is that the ones that appear first will be those with only one member and probably almost no activity.

Maybe it would be better to have a 'exclude closed teams' check box that is on by default. I think most folks going through browse are looking to find a team, although I suppose there is a use case for browsing to find a team you know about. In that case, search would be a more likely way to find a team by name.

andy-armstrong commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -516,6 +523,52 @@ def verify_on_page(self, page_num, total_teams, pagination_header_text, footer_v
+        ('last_activity_at', 'open_slots', True)
+    )
+    @ddt.unpack
+    def test_sort_teams(self, sort_order, secondary_sort_order, reverse):
```
It would be good to have a test for the default sort order too.

andy-armstrong commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -516,6 +523,52 @@ def verify_on_page(self, page_num, total_teams, pagination_header_text, footer_v
+            # Django 1.8.
+            team['last_activity_at'] = parse(team['last_activity_at']).replace(microsecond=0)
+        # Re-authenticate as staff after creating users
+        _ = AutoAuthPage(
```
Nit: we generally use underscore for i18n and double underscore (``__``) for variables we don't intend to use. In this case, why do you even need to get the user info as you aren't using the variable (or am I misunderstanding how this logic works?)

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/collections/team.js:
```diff
@@ -3,6 +3,8 @@
     define(['teams/js/collections/base', 'teams/js/models/team', 'gettext'],
         function(BaseCollection, TeamModel, gettext) {
             var TeamCollection = BaseCollection.extend({
+                sortField: 'name',
```
As a philosophical question, I'm not sure how to handle defaults like this. The server API has a default sort order already, so I think it would be better to have the default be null, and then have the parse method pick it out. Does the API return the sort order that was used? 

I see this issue with the page sizes too. We have default sizes sprinkled around both the front end and the API. I don't know whether it makes sense to have the front end just pick up the server's default, or whether it is clearer for the front-end to have its own declared defaults.

You don't have to do anything about this, but it is something I've been wondering about for a few days now.

cahrens replied:
Note that in this case we are planning to make the default sort different from what is in the API (which is alphabetical).

andy-armstrong replied:
I would argue that we should change the API too. Why would it default to a non-useful sort order?

andy-armstrong commented:
Looks great @peter-fogg. I'm done with my review, so let me know when everyone's feedback has been addressed.

cahrens commented:
Note that the secondary sort for open slots IS last_activity_at.

andy-armstrong commented:
Thanks @cahrens. I don't know how I didn't see that the secondary sort is last activity. I still think it ends up with a few pages of all the worst teams, only then followed by teams that are both active and have a few members already.

explorerleslie commented:
@andy-armstrong thanks for the feedback. The reason we implemented the "open slots" sort was because ideally we really don't want any student to end up in a single-person team. So, even if those teams are the "least popular", we also want to promote them as much as we can to so that students don't end up alone. However, I think you bring up a good point in that maybe we should sort teams with zero students last in this view, so that we don't show "Abandoned" teams first, but only teams with that one hard-working, strapping person who's trying to make a go of it. Thoughts?

andywaldrop commented:
@andy-armstrong @explorerleslie I wasn't involved in the original use cases for this, so just adding some thoughts on the "open slots" questions. I do like the idea of showing teams with available spaces upfront, as I see one of the main uses for browsing would be to find a team a user could join. This could be done as @andy-armstrong suggested with a secondary sort checkbox "show only open teams" or similar. It could also be done as a header in the results. "Open teams" could be displayed by default first with a secondary section of "Closed teams" below. That might eliminate the need for an "Open slots" sort in general. I'm a little concerned about showing teams with zero students last as they would probably never get filled.

cahrens commented:
I think it would be very odd to put teams with zero users last as the sort order would be extremely hard to decipher.

Can I request that any changes we make are for the next version of teams? :)

andywaldrop commented:
One other NIT I have is the navigation between topics and teams. When a user is hovering over the div for either the entire box takes focus, but the user can't click into the topic or team without clicking directly on the CTA arrow.

peter-fogg commented:
@andy-armstrong @bderusha I've addressed your feedback:
* Changed the cursor on the field label from a pointer to the browser's default
* made `last_activity_at` the default sort and updated Bok Choy tests
* added a test that `last_activity_at` is the default sort
* removed the unused variable from Bok Choy test

explorerleslie commented:
@andy-armstrong @cahrens let's not change anything about open slots sorting in this PR.

peter-fogg commented:
Closing this in favor of #9504.

andy-armstrong commented:
:+1:
